docs/contributing/testing_guide.rst

```diff
@@ -781,6 +463,320 @@ should be balanced with each other rather than absolute principles.
   code, consider revising the functionality being tested so that it can
   be called in a single line of code :cite:p:`khorikov:2020`.
 
+Running tests
+=============
+
+PlasmaPy's tests can be run in the following ways:
+
+1. Creating and updating a pull request on GitHub_.
+2. Running pytest_ from the command line.
+3. Running tox_ from the command line.
+4. Running tests from an :wikipedia:`integrated development environment
+   <integrated_development_environment>` (IDE).
+
+We recommend that new contributors perform the tests via a pull request
+on GitHub_. Creating a draft pull request and keeping it updated will
+ensure that the necessary checks are run frequently. This approach is
+also appropriate for pull requests with a limited scope. This advantage
+of this approach is that the tests are run automatically and do not
+require any extra work. The disadvantages are that running the tests on
+GitHub_ is often slow and that navigating the test results is sometimes
+difficult.
+
+We recommend that experienced contributors run tests either by using
+pytest_ from the command line or by using your preferred IDE.
+Using tox_ is an alternative to pytest_, but running tests with tox_
+adds the overhead of creating an isolated environment for your test and
+can thus be slower.
+
+Using GitHub
+------------
+
+The recommended way for new contributors to run PlasmaPy's full test
+suite is to `create a pull request`_ from your development branch to
+`PlasmaPy's GitHub repository`_. The test suite will be run
+automatically when the pull request is created and every time changes
+are pushed to the development branch on GitHub_. Most of these checks
+have been automated using `GitHub Actions`_.
+
+The following image shows how the results of the checks will appear in
+each pull request near the end of the *Conversation* tab. Checks that
+pass are marked with ✔️, while tests that fail are marked with ❌. Click
+on *Details* for information about why a particular check failed.
+
+.. image:: ../_static/contributor_guide/CI_checks_for_a_PR_from_2021.png
+   :width: 700
+   :align: center
+   :alt: Continuous integration test results during a pull request
+
+The following checks are performed with each pull request.
+
+* Checks with labels like **CI / Python 3.x (pull request)** verify that
+  PlasmaPy works with different versions of Python and other
+  dependencies, and on different operating systems. These tests are set
+  up using tox_ and run with pytest_ via `GitHub Actions`_. When
+  multiple tests fail, investigate these tests first.
+
+  .. tip::
+
+     `Python 3.10 <https://docs.python.org/3.10/whatsnew/3.10.html>`__,
+     `Python 3.11 <https://docs.python.org/3.11/whatsnew/3.11.html>`__,
+     and
+     `Python 3.12 <https://docs.python.org/3.12/whatsnew/3.12.html>`__
+     include (or will include) significant improvements to common error
+     messages.
```

Comment:
I updated this since there are [improved error messages in Python 3.12](https://docs.python.org/3.12/whatsnew/3.12.html#improved-error-messages) too.

Reply:
I didn't even realise 3.12 had been released